File: services/controller/api/backend_catalog.py
"""Canonical backend identities and compatibility aliases.

Product names, persisted IDs, and transport-era aliases are deliberately kept
separate here. New state is always written with the canonical ID; old state is
accepted at the boundary and normalized before it reaches application logic.
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class BackendDefinition:
    id: str
    label: str


JAEGER_BACKEND_ID = "jaeger_local"

BACKENDS: tuple[BackendDefinition, ...] = (
    BackendDefinition("hermes_local", "Hermes Agent"),
    BackendDefinition(JAEGER_BACKEND_ID, "Jaeger AI"),
    BackendDefinition("claude_local", "Claude Code"),
    BackendDefinition("codex_local", "OpenAI Codex"),
    BackendDefinition("gemini_local", "Google Gemini"),
    BackendDefinition("grok_local", "xAI Grok"),
    BackendDefinition("opencode_local", "OpenCode"),
    BackendDefinition("cursor_local", "Cursor"),
    BackendDefinition("pi_local", "Pi Coding Agent"),
    BackendDefinition("openai_cloud", "OpenAI"),
    BackendDefinition("xai_cloud", "xAI Grok"),
    BackendDefinition("gemini_cloud", "Google Gemini API"),
    BackendDefinition("gemini_antigravity", "Gemini (Antigravity IDE)"),
    BackendDefinition("ollama_local", "Ollama"),
)

BACKEND_BY_ID = {backend.id: backend for backend in BACKENDS}
VALID_BACKEND_IDS = tuple(BACKEND_BY_ID)

# Read-only migration boundary. Never expose or persist these values as the
# selected backend after normalization.
BACKEND_ALIASES = {
    "hermes": "hermes_local",
    "jaeger": JAEGER_BACKEND_ID,
    "jaegerai": JAEGER_BACKEND_ID,
    "jaeger_ai": JAEGER_BACKEND_ID,
    "jros": JAEGER_BACKEND_ID,
    "jros_local": JAEGER_BACKEND_ID,
}


def normalize_backend_id(value: object, *, fallback: str = "") -> str:
    raw = str(value or "").strip().lower()
    normalized = BACKEND_ALIASES.get(raw, raw)
    if normalized in BACKEND_BY_ID:
        return normalized
    normalized_fallback = BACKEND_ALIASES.get(str(fallback or "").strip().lower(), fallback)
    return normalized_fallback if normalized_fallback in BACKEND_BY_ID else ""


def backend_display_name(value: object) -> str:
    normalized = normalize_backend_id(value)
    if not normalized:
        return str(value or "")
    return BACKEND_BY_ID[normalized].label

File: apps/web/src/shared/backend-catalog.ts
export const JAEGER_BACKEND_ID = "jaeger_local";

export interface BackendMeta {
  label: string;
  color: string;
}

// Must stay in sync with services/controller/api/backend_catalog.py::BACKEND_META
// See services/controller/tests/test_backend_catalog_ts_parity.py
export const BACKEND_META: Readonly<Record<string, BackendMeta>> = {
  hermes_local:   { label: "Hermes Agent", color: "#08EBF1" },
  jaeger_local:   { label: "Jaeger AI", color: "#3889FD" },
  claude_local:   { label: "Claude Code", color: "#D97706" },
  codex_local:    { label: "OpenAI Codex", color: "#10B981" },
  gemini_local:   { label: "Google Gemini", color: "#6366F1" },
  gemini_cloud:   { label: "Google Gemini API", color: "#6366F1" },
  gemini_antigravity: { label: "Gemini (Antigravity IDE)", color: "#6366F1" },
  grok_local:     { label: "xAI Grok", color: "#8B5CF6" },
  opencode_local: { label: "OpenCode", color: "#EC4899" },
  cursor_local:   { label: "Cursor", color: "#06B6D4" },
  ollama_local:   { label: "Ollama", color: "#F59E0B" },
  openai_cloud:   { label: "OpenAI", color: "#10A37F" },
  xai_cloud:      { label: "xAI Grok", color: "#8B5CF6" },
  pi_local:       { label: "Pi Coding Agent", color: "#F472B6" },
};

const BACKEND_ALIASES: Readonly<Record<string, string>> = {
  jaeger: JAEGER_BACKEND_ID,
  jaegerai: JAEGER_BACKEND_ID,
  jaeger_ai: JAEGER_BACKEND_ID,
  jros: JAEGER_BACKEND_ID,
  jros_local: JAEGER_BACKEND_ID,
  hermes: "hermes_local",
};

/** Normalize API, imported-session, and saved legacy values at the UI edge. */
export function normalizeBackendId(value: unknown): string {
  const raw = String(value ?? "").trim().toLowerCase();
  return BACKEND_ALIASES[raw] ?? raw;
}

export function backendLabel(value: unknown): string {
  const id = normalizeBackendId(value);
  return BACKEND_META[id]?.label
    ?? id.replace(/_/g, " ").replace(/\b\w/g, (character) => character.toUpperCase());
}

export function backendColor(value: unknown): string {
  return BACKEND_META[normalizeBackendId(value)]?.color ?? "#6b7194";
}
